src/main/java/com/google/devtools/build/lib/syntax/MethodDescriptor.java

```diff
@@ -59,19 +62,157 @@ private MethodDescriptor(
       boolean allowReturnNones,
       boolean useStarlarkThread,
       boolean useStarlarkSemantics) {
+    boolean isStringModule = method.getDeclaringClass() == StringModule.class;
+
     this.method = method;
     this.annotation = annotation;
     this.name = name;
     this.doc = doc;
     this.documented = documented;
     this.structField = structField;
-    this.parameters = parameters;
+    this.parameters =
+        !isStringModule ? parameters : Arrays.copyOfRange(parameters, 1, parameters.length);
     this.extraPositionals = extraPositionals;
     this.extraKeywords = extraKeywords;
     this.selfCall = selfCall;
     this.allowReturnNones = allowReturnNones;
     this.useStarlarkThread = useStarlarkThread;
     this.useStarlarkSemantics = useStarlarkSemantics;
+    try {
+      MethodHandle mh = MethodHandles.lookup().unreflect(method);
+
+      if (isStringModule) {
+        // String methods get the string as an extra argument
+        // because their true receiver is StringModule.INSTANCE.
+        mh = MethodHandles.insertArguments(mh, 0, StringModule.INSTANCE);
+      }
+
+      if (!useStarlarkThread) {
```

Comment:
Comments at every step of the way would help here.

Reply:
Done.